lib/contractWrapperFactory.ts

```diff
@@ -49,9 +88,50 @@ export class ContractWrapperFactory<TWrapper extends IContractWrapperBase>
    */
   public async deployed(): Promise<TWrapper> {
     await this.ensureSolidityContract();
-    return new this.wrapper(this.solidityContract, this.web3EventService).hydrateFromDeployed();
+
+    const getWrapper = (): Promise<TWrapper> => {
+      return new this.wrapper(this.solidityContract, this.web3EventService).hydrateFromDeployed();
+    };
+
+    let hydratedWrapper: TWrapper;
+
+    if (ContractWrapperFactory.configService.get("cacheContractWrappers")) {
+      hydratedWrapper = this.getCachedContract(this.solidityContractName) as TWrapper;
+      if (!hydratedWrapper) {
+        hydratedWrapper = await getWrapper();
+        // this'll cache for both `.at` and `.deployed`
+        this.setCachedContract(this.solidityContractName, hydratedWrapper);
+      }
+    } else {
+      hydratedWrapper = await getWrapper();
+    }
+    return hydratedWrapper;
+  }
+
+  private getCachedContract(name: string, at?: string): IContractWrapperBase | undefined {
+    const addressMap = ContractWrapperFactory.contractCache.get(name);
+    if (!addressMap) {
+      return undefined;
+    }
+    return addressMap.get(at ? at : `${name}_deployed`);
   }
 
+  private setCachedContract(
+    name: string,
+    wrapper: IContractWrapperBase,
+    at?: string): void {
+
+    let addressMap = ContractWrapperFactory.contractCache.get(name);
+    if (!addressMap) {
+      addressMap = new Map<Address, IContractWrapperBase>();
+      ContractWrapperFactory.contractCache.set(name, addressMap);
+    }
+    addressMap.set(wrapper.address, wrapper);
+    if (!at) {
+      // so we can also get the deployed version of this contract without the address
```

Comment:
The "${name}_deployed" key is for when users call `deployed()`, which supplies no address.

Reply:
right, but the deployed contract has an address, which you could use as the name of the key so that you dont end up with two copies of the same cached contract

Reply:
Found a way to do it